File: abg.py
# acsii-based graphics format

file = open(input("open a file... "), "r")

class abg:
	def interpret(file):
		try:
			lines = file.read().splitlines()

			if file.name.lower().endswith(".abg"):
				brightnesslevels = ["1", "2", "3", "4", "x"]
				brightness = ["  ", "░░", "▓▓", "██", "\033[0m  \033[40m"]
				colornames = ["r", "g", "y", "c", "p", "b", "w", "x"]
				colors = ["\033[31m", "\033[32m", "\033[33m", "\033[34m", "\033[35m", "\033[36m", "\033[37m", ""]
				result = "\033[40m"
				for y in range(len(lines)-1):
					linepixels = lines[y+1].split(" ")
					for x in range(len(linepixels)):
						pixel = linepixels[x]
						if pixel[0].lower() in colornames:
							result += colors[colornames.index(pixel[0].lower())]
							if pixel[1].lower() in brightnesslevels:
								result += brightness[brightnesslevels.index(pixel[1].lower())]
							else:
								return "invalid brightness level"
						else:
							return "invalid color"
					result += "\n"
				return result
			else:
				return "this is not a .abg file"
		except AttributeError:
			return "something went wrong!"

print(abg.interpret(file))

File: abg.rs
pub fn read(file: String) {
	let lines: Vec<_> = file.split("\n").collect();
	if lines[0] == "1.0.1" {
		let brightnesslevels = vec!["1", "2", "3", "4", "x"];
		let brightness = ["  ", "░░", "▓▓", "██", "\x1b[0m  \x1b[40m"];
		let colornames = vec!["r", "g", "y", "v", "p", "b", "w", "x"];
		let colors = ["\x1b[31m", "\x1b[32m", "\x1b[33m", "\x1b[34m", "\x1b[35m", "\x1b[36m", "\x1b[37m", ""];
		for y in 0..lines.len()-2 {
			let linepixels: Vec<_> = lines[y+2].split(" ").collect();
			for x in 0..linepixels.len()-1 {
				let pixel: Vec<char> = linepixels[x].chars().collect();
				if colornames.contains(&pixel[0].to_string().as_str()) {
					print!("{}", colors[colornames.iter().position(|&r| r == pixel[0].to_string()).unwrap()]);
				} else {
					print!("\x1b[37m");
				}
				if brightnesslevels.contains(&pixel[1].to_string().as_str()) {
					print!("{}", brightness[brightnesslevels.iter().position(|&r| r == pixel[1].to_string()).unwrap()]);
				} else {
					print!("??");
				}
			}
			println!("");
		}
	}
}